src/toxcore/dht_new/packet.rs

```diff
@@ -0,0 +1,1039 @@
+    /// UDP over IPv6.
+    U6 = 10,
+    /// TCP over IPv4.
+    T4 = 130,
+    /// TCP over IPv6.
+    T6 = 138,
+}
+
+/// Match first byte from the provided slice as `IpType`. If no match found,
+/// return `Error`.
+impl FromBytes for IpType {
+    named!(from_bytes<IpType>, switch!(le_u8,
+        2   => value!(IpType::U4) |
+        10  => value!(IpType::U6) |
+        130 => value!(IpType::T4) |
+        138 => value!(IpType::T6)
+    ));
+}
+
+// TODO: move it somewhere else
+impl ToBytes for IpAddr {
+    fn to_bytes<'a>(&self, buf: (&'a mut [u8], usize)) -> Result<(&'a mut [u8], usize), GenError> {
+        match *self {
+            IpAddr::V4(ref p) => p.to_bytes(buf),
+            IpAddr::V6(ref p) => p.to_bytes(buf),
+        }
+    }
+}
+
+// TODO: move it somewhere else
+/// Fail if there are less than 4 bytes supplied, otherwise parses first
+/// 4 bytes as an `Ipv4Addr`.
+impl FromBytes for Ipv4Addr {
+    named!(from_bytes<Ipv4Addr>, map!(count!(le_u8, 4), 
+        |v| Ipv4Addr::new(v[0], v[1], v[2], v[3])
+    ));
+}
+
+impl ToBytes for Ipv4Addr {
+    fn to_bytes<'a>(&self, buf: (&'a mut [u8], usize)) -> Result<(&'a mut [u8], usize), GenError> {
+        let o = self.octets();
+        do_gen!(buf,
+            gen_be_u8!(o[0]) >>
+            gen_be_u8!(o[1]) >>
+            gen_be_u8!(o[2]) >>
+            gen_be_u8!(o[3]) 
+        )
+    }
+}
+
+// TODO: move it somewhere else
+/// Fail if there are less than 16 bytes supplied, otherwise parses first
+/// 16 bytes as an `Ipv6Addr`.
+impl FromBytes for Ipv6Addr {
+    named!(from_bytes<Ipv6Addr>, map!(count!(le_u16, 8), 
+        |v| Ipv6Addr::new(v[0], v[1], v[2], v[3], v[4], v[5], v[6], v[7])
+    ));
+}
+
+impl ToBytes for Ipv6Addr {
+    fn to_bytes<'a>(&self, buf: (&'a mut [u8], usize)) -> Result<(&'a mut [u8], usize), GenError> {
+        let s = self.segments();
+        do_gen!(buf,
+            gen_le_u16!(s[0]) >>
+            gen_le_u16!(s[1]) >>
+            gen_le_u16!(s[2]) >>
+            gen_le_u16!(s[3]) >>
+            gen_le_u16!(s[4]) >>
+            gen_le_u16!(s[5]) >>
+            gen_le_u16!(s[6]) >>
+            gen_le_u16!(s[7]) 
+        )
+    }
+}
+
+/** `PackedNode` format is a way to store the node info in a small yet easy to
+parse format.
+
+It is used in many places in Tox, e.g. in `DHT Send nodes`.
+
+To store more than one node, simply append another on to the previous one:
+
+`[packed node 1][packed node 2][...]`
+
+Serialized Packed node:
+
+Length | Content
+------ | -------
+`1`    | [`IpType`](./.enum.IpType.html)
+`4` or `16` | IPv4 or IPv6 address
+`2`    | port
+`32`   | node ID
+
+Size of serialized `PackedNode` is 39 bytes with IPv4 node info, or 51 with
+IPv6 node info.
+
+DHT module *should* use only UDP variants of `IpType`, given that DHT runs
+solely on the UDP.
+*/
+#[derive(Copy, Clone, Debug, Eq, PartialEq)]
+pub struct PackedNode {
+    /// IP type, includes also info about protocol used.
+    ip_type: IpType,
```

Comment:
https://zetok.github.io/tox-spec/#dht-operation

> Only the UDP Protocol (IP Type 2 and 10) are used in the DHT module when sending nodes with the packed node format.

So you don't have to keep ip_type here, because it is either UDP4 or UDP6, but the info about 4 or 6 is stored inside SocketAddr, since it is:

```
pub enum SocketAddr {
    V4(SocketAddrV4),
    V6(SocketAddrV6),
}

So you would add a method to serialize/deserialize SocketAddr and based on V4 or V6 you have to add tag U4 or U6 or check that tag.
```